odoo/addons/base/models/ir_qweb.py

```diff
@@ -105,25 +186,159 @@ def _render(self, template, values=None, **options):
 
     # apply ormcache_context decorator unless in dev mode...
     @tools.conditional(
         'xml' not in tools.config['dev_mode'],
-        tools.ormcache('id_or_xml_id', 'tuple(options.get(k) for k in self._get_template_cache_keys())'),
+        tools.ormcache('template', 'tuple(options.get(k) for k in self._get_template_cache_keys())'),
     )
     @QwebTracker.wrap_compile
-    def _compile(self, id_or_xml_id, options):
+    def _compile(self, template, options):
+        """ Compile the given template into a rendering function (generator)::
+
+            render(qweb, values)
+
+        where ``qweb`` is a QWeb instance and ``values`` are the values to render.
+        """
+        if options is None:
+            options = {}
+
+        element, document, ref = self._get_template(template, options)
+        if not ref:
+            ref = element.get('t-name', str(document))
+
+        # reference to get xml and etree (usually the template ID)
+        options['ref'] = ref
+        # str xml of the reference template used for compilation. Useful for debugging, dev mode and profiling.
+        options['ref_xml'] = document if isinstance(document, str) else str(document, 'utf-8')
+
+        _options = dict(options)
+        options = frozendict(options)
+
+        # Initial template value send to render method (not in the froozen dict because it may be
+        # different from one render to another. Indeed, it may be the view ID or the key)
+        _options['template'] = template
+        # Root of the etree which will be processed during compilation.
+        _options['root'] = element.getroottree()
+        # Reference to the last node being compiled. It is mainly used for debugging and displaying
+        # error messages.
+        _options['last_path_node'] = None
+
+        if not options.get('nsmap'):
+            _options['nsmap'] = {}
+
+        # generate code
+
+        def_name = f"template_{ref}" if isinstance(ref, int) else "template"
+
         try:
-            id_or_xml_id = int(id_or_xml_id)
+            _options['_text_concat'] = []
+            self._appendText("", _options) # To ensure the template function is a generator and doesn't become a regular function
+            code_lines = ([f"def {def_name}(self, values, log):"] +
+                self._compile_node(element, _options, 1) +
+                self._flushText(_options, 1))
+        except QWebException as e:
+            raise e
+        except Exception as e:
+            raise QWebException("Error when compiling xml template",
+                self, options, e, template, path=_options.get('last_path_node')) from e
+        try:
+            code = '\n'.join(code_lines)
+        except QWebException as e:
+            raise e
+        except Exception as e:
+            code = '\n'.join(map(str, code_lines))
+            raise QWebException("Error when compiling xml template",
+                self, options, e, template, code=code) from e
```

Comment:
Sorry if there is something I don't see, but how can line 252 raise errors ?

Reply:
If we forget in a widget or directive to return a string and for example we put a float.